### PR TITLE
WIP - Fix bug  JSON input barfs on {"emptylist":[]} 

### LINKED ISSUE #1036: JSON input barfs on {"emptylist":[]}
**Describe the bug**
JSON inuput barfs on {"emptylist":[]}

```bash
stack backtrace:
   0: rust_begin_unwind
             at /rustc/0b42deaccc2cbe17a68067aa5fdb76104369e1fd/library/std/src/panicking.rs:498:5
   1: core::panicking::panic_fmt
             at /rustc/0b42deaccc2cbe17a68067aa5fdb76104369e1fd/library/core/src/panicking.rs:107:14
   2: parquet::arrow::levels::LevelInfo::filter_array_indices
   3: parquet::arrow::arrow_writer::write_leaf
   4: parquet::arrow::arrow_writer::write_leaves
   5: parquet::arrow::arrow_writer::write_leaves
   6: parquet::arrow::arrow_writer::ArrowWriter<W>::write
   7: json2parquet::main
```

I was driving the library with https://github.com/domoritz/json2parquet

```bash
thread 'main' panicked at 'Cannot filter indices on a non-primitive array, found List(true)', /PATH/parquet-6.1.0/src/arrow/levels.rs:757:18
```

**To Reproduce**
{"emptylist":[]}

**Expected behavior**
Same as pyarrow, it does not barf the Parquet writer.

```bash
pyarrow.Table
emptylist: list<item: null>
  child 0, item: null
----
emptylist: [[0 nulls]]
```

**Additional context**

### PR DESCRIPTION
**Pull request by novemberkilo**
# Which issue does this PR close?

Closes #1036 

# Rationale for this change
 
 

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: parquet/src/arrow/arrow_writer.rs [hunk 1]
```diff
@@ -1546,6 +1546,38 @@ mod tests {
         one_column_roundtrip("list_single_column", values, true, Some(SMALL_SIZE / 2));
     }
 
+    #[test]
```
Comment (alamb) [inline parquet/src/arrow/arrow_writer.rs:1549]: 💯  for test driven development 👍

Comment (novemberkilo) [reply parquet/src/arrow/arrow_writer.rs:1549]: Thanks :)

Also pls https://github.com/apache/arrow-rs/issues/1036#issuecomment-997521410

File: parquet/src/arrow/levels.rs [hunk 2]
```diff
@@ -751,9 +751,10 @@ impl LevelInfo {
 
     /// Given a level's information, calculate the offsets required to index an array correctly.
     pub(crate) fn filter_array_indices(&self) -> Vec<usize> {
-        // happy path if not dealing with lists
+        dbg!(&self);
         let is_nullable = match self.level_type {
             LevelType::Primitive(is_nullable) => is_nullable,
+            LevelType::List(is_nullable) => is_nullable,
```
Comment (novemberkilo) [inline parquet/src/arrow/levels.rs:757]: This is the only change to the code path -- it addresses the error message we were seeing where the `List` leveltype was being picked up by the catchall (thus the panic).

File: parquet/src/arrow/arrow_writer.rs [hunk 3]
```diff
@@ -1574,6 +1570,9 @@ mod tests {
         .unwrap();
 
         let a = ListArray::from(a_list_data);
+        // let values = Arc::new(a);
+        // one_column_roundtrip("null_list_single_column", values, true, Some(SMALL_SIZE / 2));
+
         let batch = RecordBatch::try_new(Arc::new(schema), vec![Arc::new(a)]).unwrap();
         roundtrip("test_null_list_single_column.parquet", batch, None);
```
Comment (novemberkilo) [inline parquet/src/arrow/arrow_writer.rs:1577]: The error I am now seeing comes from the reader being unable to read the parquet that is written out. 

```
thread 'arrow::arrow_writer::tests::null_list_single_column' panicked at 'called `Result::unwrap()` on an `Err` value: ArrowError("Unable to convert parquet INT32 logical type Some(UNKNOWN(NullType)) or converted type NONE")', parquet/src/arrow/array_reader.rs:1502:14

```

I have peppered the code with `dbg!` to see if I can make sense of what's going on but so far all I have is a hunch that we've also got a potential bug in the reader. Detailed output from this test appears below:

```
---- arrow::arrow_writer::tests::null_list_single_column stdout ----
[parquet/src/arrow/levels.rs:754] &self = LevelInfo {
    definition: [
        1,
    ],
    repetition: Some(
        [
            0,
        ],
    ),
    array_offsets: [
        0,
        0,
    ],
    array_mask: [
        true,
    ],
    max_definition: 2,
    level_type: List(
        true,
    ),
    offset: 0,
    length: 0,
}
[parquet/src/arrow/levels.rs:788] &filtered = []
[parquet/src/arrow/array_reader.rs:1844] &context = ArrayReaderBuilderContext {
    def_level: 0,
    rep_level: 0,
    path: ColumnPath {
        parts: [
            "arrow_schema",
        ],
    },
}
[parquet/src/arrow/array_reader.rs:1845] &child = GroupType {
    basic_info: BasicTypeInfo {
        name: "emptylist",
        repetition: Some(
            OPTIONAL,
        ),
        converted_type: LIST,
        logical_type: Some(
            LIST(
                ListType,
            ),
        ),
        id: None,
    },
    fields: [
        GroupType {
            basic_info: BasicTypeInfo {
                name: "list",
                repetition: Some(
                    REPEATED,
                ),
                converted_type: NONE,
                logical_type: None,
                id: None,
            },
            fields: [
                PrimitiveType {
                    basic_info: BasicTypeInfo {
                        name: "item",
                        repetition: Some(
                            OPTIONAL,
                        ),
                        converted_type: NONE,
                        logical_type: Some(
                            UNKNOWN(
                                NullType,
                            ),
                        ),
                        id: None,
                    },
                    physical_type: INT32,
                    type_length: -1,
                    scale: -1,
                    precision: -1,
                },
            ],
        },
    ],
}
[parquet/src/arrow/array_reader.rs:1495] &list_type = GroupType {
    basic_info: BasicTypeInfo {
        name: "emptylist",
        repetition: Some(
            OPTIONAL,
        ),
        converted_type: LIST,
        logical_type: Some(
            LIST(
                ListType,
            ),
        ),
        id: None,
    },
    fields: [
        GroupType {
            basic_info: BasicTypeInfo {
                name: "list",
                repetition: Some(
                    REPEATED,
                ),
                converted_type: NONE,
                logical_type: None,
                id: None,
            },
            fields: [
                PrimitiveType {
                    basic_info: BasicTypeInfo {
                        name: "item",
                        repetition: Some(
                            OPTIONAL,
                        ),
                        converted_type: NONE,
                        logical_type: Some(
                            UNKNOWN(
                                NullType,
                            ),
                        ),
                        id: None,
                    },
                    physical_type: INT32,
                    type_length: -1,
                    scale: -1,
                    precision: -1,
                },
            ],
        },
    ],
}
[parquet/src/arrow/array_reader.rs:1496] &context = ArrayReaderBuilderContext {
    def_level: 0,
    rep_level: 0,
    path: ColumnPath {
        parts: [
            "arrow_schema",
        ],
    },
}
[parquet/src/arrow/array_reader.rs:1497] &item_type = PrimitiveType {
    basic_info: BasicTypeInfo {
        name: "item",
        repetition: Some(
            OPTIONAL,
        ),
        converted_type: NONE,
        logical_type: Some(
            UNKNOWN(
                NullType,
            ),
        ),
        id: None,
    },
    physical_type: INT32,
    type_length: -1,
    scale: -1,
    precision: -1,
}
[parquet/src/arrow/array_reader.rs:1498] &new_context = ArrayReaderBuilderContext {
    def_level: 2,
    rep_level: 1,
    path: ColumnPath {
        parts: [
            "arrow_schema",
            "emptylist",
        ],
    },
}
thread 'arrow::arrow_writer::tests::null_list_single_column' panicked at 'called `Result::unwrap()` on an `Err` value: ArrowError("Unable to convert parquet INT32 logical type Some(UNKNOWN(NullType)) or converted type NONE")', parquet/src/arrow/array_reader.rs:1502:14

```

**Comment by novemberkilo** [comment]
@alamb FYI I am afk until early Jan now so won't be making any progress on this until then.

**Comment by novemberkilo** [comment]
Still here and interested. Sorry for the hiatus - will try and spin back up shortly. Confirming that this is still relevant pls @alamb

**Comment by novemberkilo** [comment]
@alamb I have an update.

1. I have confirmed that the issue persists as of current `master` 
2. I have re-examined and confirmed the approach -- I think that the test here is correct
3. I've built `json2parquet` with the fix that appears in this PR (see [this comment](https://github.com/apache/arrow-rs/pull/1063#pullrequestreview-838891472) specifically) and this results in a parquet file. 
4. I can use `pyarrow` to read the file written out in step 3. 

I am not sure whether I have uncovered a bug with `array_reader` or if I am writing out a dud parquet file. I want to investigate further and if I am reasonably sure that I have uncovered a bug, I will create an issue with a minimal example to reproduce.

I am going to close this PR for now because it is stale and a bit noisy. I will recreate a clean PR when I am clearer about the outcome of the code that appears here.

Copying @tustvold because I wonder if this is now close to the work you've done recently in https://github.com/apache/arrow-rs/pull/1246

**Comment by alamb** [comment]
Thanks @novemberkilo

**Pull request closed without merging**